### PR TITLE
Fix for ExtendedBlockStates containing at least one IProperty and one IUnlistedProperty preventing block placement.

### PR DESCRIPTION
**Pull request by shadekiller666**

### CODE REVIEW

File: patches/minecraft/net/minecraft/block/state/BlockState.java.patch [hunk 1]
```diff
@@ -1,6 +1,46 @@
...
+-import com.google.common.base.Joiner;
+-import com.google.common.base.Objects;
+-import com.google.common.collect.HashBasedTable;
+-import com.google.common.collect.ImmutableList;
+-import com.google.common.collect.ImmutableMap;
+-import com.google.common.collect.ImmutableTable;
+-import com.google.common.collect.Iterables;
+-import com.google.common.collect.Lists;
+-import com.google.common.collect.Maps;
+ import java.util.ArrayList;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/state/BlockState.java.patch]: no....

File: patches/minecraft/net/minecraft/block/state/BlockStateBase.java.patch [hunk 2]
```diff
@@ -0,0 +1,17 @@
+--- ../src-base/minecraft/net/minecraft/block/state/BlockStateBase.java
++++ ../src-work/minecraft/net/minecraft/block/state/BlockStateBase.java
+@@ -2,6 +2,7 @@
+ 
+ import com.google.common.base.Function;
+ import com.google.common.base.Joiner;
++import com.google.common.collect.ImmutableTable;
+ import com.google.common.collect.Iterables;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/state/BlockStateBase.java.patch]: no...

File: patches/minecraft/net/minecraft/block/state/BlockStateBase.java.patch [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ import java.util.Collection;
+ import java.util.Iterator;
+@@ -76,4 +77,6 @@
+ 
+         return stringbuilder.toString();
+     }
++    
++    public abstract ImmutableTable<IProperty, Comparable, IBlockState> getPropertyValueTable();
+ }
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/state/BlockStateBase.java.patch]: for compatiilities sake this needs to be non-abstract and probably null return.

File: patches/minecraft/net/minecraft/client/renderer/BlockModelShapes.java.patch [hunk 4]
```diff
@@ -1,6 +1,27 @@
...
++
+ import net.minecraft.block.Block;
+ import net.minecraft.block.BlockBed;
+ import net.minecraft.block.BlockCactus;
+@@ -57,6 +59,7 @@
+ import net.minecraft.init.Blocks;
+ import net.minecraft.util.EnumFacing;
+ import net.minecraft.util.ResourceLocation;
++import net.minecraftforge.fml.common.FMLLog;
+ import net.minecraftforge.fml.relauncher.Side;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/client/renderer/BlockModelShapes.java.patch]: Wat

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 5]
```diff
@@ -1,11 +1,37 @@
...
+ import net.minecraft.block.Block;
+ import net.minecraft.block.BlockHopper;
+ import net.minecraft.block.BlockLiquid;
+@@ -19,6 +21,7 @@
+ import net.minecraft.block.BlockSnow;
+ import net.minecraft.block.BlockStairs;
+ import net.minecraft.block.material.Material;
++import net.minecraft.block.state.BlockState;
+ import net.minecraft.block.state.IBlockState;
+ import net.minecraft.command.IEntitySelector;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Again

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 6]
```diff
@@ -99,8 +126,12 @@
      }
  
      public boolean func_175667_e(BlockPos p_175667_1_)
-@@ -286,39 +318,58 @@
+@@ -284,41 +321,62 @@
+         }
+         else
          {
++            if (p_180501_2_ instanceof net.minecraftforge.common.property.IExtendedBlockState)
++                p_180501_2_ = ((net.minecraftforge.common.property.IExtendedBlockState) p_180501_2_).getClean();
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/World.java.patch]: this is probably only needed in chunk itself, when its being directly added to the array.

File: patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch [hunk 7]
```diff
@@ -1,16 +1,36 @@
...
++
+ import java.util.Arrays;
+ import java.util.Iterator;
+ import java.util.List;
+@@ -10,6 +11,7 @@
+ import java.util.Random;
+ import java.util.concurrent.Callable;
+ import java.util.concurrent.ConcurrentLinkedQueue;
++
+ import net.minecraft.block.Block;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch]: Tisk Tisk Tisk

File: patches/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java.patch [hunk 8]
```diff
@@ -0,0 +1,11 @@
+--- ../src-base/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
++++ ../src-work/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
+@@ -35,6 +35,8 @@
+ 
+     public void func_177484_a(int p_177484_1_, int p_177484_2_, int p_177484_3_, IBlockState p_177484_4_)
+     {
++        if (p_177484_4_ instanceof net.minecraftforge.common.property.IExtendedBlockState)
++            p_177484_4_ = ((net.minecraftforge.common.property.IExtendedBlockState) p_177484_4_).getClean();
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java.patch]: Actually this is probably the only place it's needed as this is the actual class that stores the values and what gets asked for the ids to save to the world

File: patches/minecraft/net/minecraft/world/gen/feature/WorldGenDungeons.java.patch [hunk 9]
```diff
@@ -23,7 +23,7 @@
 +                                        WeightedRandomChestContent.func_177630_a(p_180709_2_, ChestGenHooks.getItems(DUNGEON_CHEST, p_180709_2_), (TileEntityChest)tileentity1, ChestGenHooks.getCount(DUNGEON_CHEST, p_180709_2_));
                                      }
  
-                                     break label197;
+                                     break label100;
 @@ -184,6 +186,12 @@
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/gen/feature/WorldGenDungeons.java.patch]: Decompiler difference skip this patch.

File: patches/minecraft/net/minecraftforge/fml/common/registry/GameData.java.patch [hunk 10]
```diff
@@ -0,0 +1,11 @@
+--- ../src-base/minecraft/net/minecraftforge/fml/common/registry/GameData.java
++++ ../src-work/minecraft/net/minecraftforge/fml/common/registry/GameData.java
+@@ -837,6 +837,8 @@
+ 
+         for (IBlockState state : ((List<IBlockState>)block.func_176194_O().func_177619_a()))
+         {
++            if (state instanceof net.minecraftforge.common.property.IExtendedBlockState)
++                state = ((net.minecraftforge.common.property.IExtendedBlockState) state).getClean();
```
Comment (LexManos) [inline patches/minecraft/net/minecraftforge/fml/common/registry/GameData.java.patch]: In theory get listed states should ALWAYS return the clean states, but I guess it doesn't hurt to be safe.
But this needs to not be a patch it needs to be in the FML code base itself.

File: src/main/java/net/minecraftforge/client/model/ModelFluid.java [hunk 11]
```diff
@@ -139,8 +139,8 @@ public BakedFluid(TRSRTransformation transformation, VertexFormat format, int co
                         y[i] = state.getValue(BlockFluidBase.LEVEL_CORNERS[i]);
                     }
                 }
-
-                float flow = state.getValue(BlockFluidBase.FLOW_DIRECTION);
+                
+                float flow = (Float)state.getValue(BlockFluidBase.FLOW_DIRECTION);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/client/model/ModelFluid.java]: Shouldnt be needed...?

File: src/test/resources/assets/forgedebugmodelloaderregistry/blockstates/CustomModelBlock.json [hunk 12]
```diff
@@ -9,9 +9,9 @@
      },
     "variants": {
         "normal": [{
-            "transform": {
+            /*"transform": {
                 "rotation": {"y": 45}
-            }
+            }*/
```
Comment (LexManos) [inline src/test/resources/assets/forgedebugmodelloaderregistry/blockstates/CustomModelBlock.json]: Why were these commented out in favor of manually setting the rotations?

Comment (shadekiller666) [inline src/test/resources/assets/forgedebugmodelloaderregistry/blockstates/CustomModelBlock.json]: that statement made the model always be rotated to 45 degrees, and i wanted to show that the rotation based on the PropertyDirection was working properly without that extra 45 degrees

**Pull request merged**